File: Week12/hw12_var2_jacobi.py
import numpy as np
from numpy import (pi, exp, sin, cos)
import matplotlib.pyplot as plt
from time import perf_counter
import os
import winsound


# 2D Navier-Stokes equation
# Explicit scheme + Jacobi
# Problem 2

# Initial condition
# U(t=0, x, y) = 0
# V(t=0, x, y) = 0
# P(t=0, x, y) = 0

# Boundary conditions
# U(t, x=0, y) = 0
# U(t, x=1, y) = 0
# Uy(t, 0<x<0.2, y=0) = 0
# U(t, 0.2<x<0.8, y=0) = 0
# Uy(t, 0.8<x<1, y=0) = 0
# U(t, 0<x<0.3, y=1) = 0
# U(t, 0.3<x<0.7, y=1) = 0
# U(t, 0.7<x<1, y=1) = 0

# V(t, x=0, y) = 0
# V(t, x=1, y) = 0
# Vy(t, 0<x<0.2, y=0) = 0
# V(t, 0.2<x<0.8, y=0) = 0
# Vy(t, 0.8<x<1, y=0) = 0
# V(t, 0<x<0.3, y=1) = 0
# V(t, 0.3<x<0.7, y=1) = -1
# V(t, 0.7<x<1, y=1) = 0

# Px(t, x=0, y) = 0
# Px(t, x=1, y) = 0
# P(t, 0<x<0.2, y=0) = 0
# Py(t, 0.2<x<0.8, y=0) = 0
# P(t, 0.8<x<1, y=0) = 0
# Py(t, 0<x<0.3, y=1) = 0
# P(t, 0.3<x<0.7, y=1) = 1
# Py(t, 0.7<x<1, y=1) = 0

# Returns numpy array
def Explicit_Burger(S:np.ndarray, U:np.ndarray, V:np.ndarray, N, M, dx, dy, dt, nu) -> np.ndarray:
    """Explicit Euler method for solving 2D Burger's equation"""
    # Reference
    U_old = U
    V_old = V
    S_old = S
    
    S_new = np.zeros_like(S)
    
    S_new[1:M-1, 1:N-1] = S_old[1:M-1, 1:N-1] + dt*(
        - U_old[1:M-1, 1:N-1]*(S_old[1:M-1, 2:N] - S_old[1:M-1, 0:N-2]) \
                                / (2*dx) \
        - V_old[1:M-1, 1:N-1]*(S_old[2:M, 1:N-1] - S_old[0:M-2, 1:N-1]) \
                                / (2*dy) \
        + nu*(
            (S_old[1:M-1, 2:N] - 2*S_old[1:M-1, 1:N-1] + S_old[1:M-1, 0:N-2]) \
                / dx**2 \
            + (S_old[2:M, 1:N-1] - 2*S_old[1:M-1, 1:N-1] + S_old[0:M-2, 1:N-1]) \
                / dy**2))
    
    return S_new

def Jacobi_method(P:np.ndarray, U:np.ndarray, V:np.ndarray, N, M, dx, dy, dt, rho) -> np.ndarray:
    """Jacobi method for solving 2D Poisson equation"""
    # Reference
    P_old = P
    U_new = U
    V_new = V

    P_new = np.zeros_like(P)

    P_new[1:M-1, 1:N-1] = (dy**2*(P_old[1:M-1, 2:N] + P_old[1:M-1, 0:N-2]) \
        + dx**2*(P_old[2:M, 1:N-1] + P_old[0:M-2, 1:N-1])) \
            / (2*(dx**2 + dy**2)) \
        - dx**2*dy**2*rho \
            / (2*dt*(dx**2 + dy**2)) \
        * ((U_new[1:M-1, 2:N] - U_new[1:M-1, 0:N-2]) \
            / (2*dx) \
        + (V_new[2:M, 1:N-1] - V_new[0:M-2, 1:N-1]) \
            / (2*dy))
    
    return P_new

# Void function
# Reference
def set_boundary_U(U:np.ndarray):
    U[0:M, 0] = 0
    U[0:M, N-1] = 0
    U[0, 0:N1] = U[1, 0:N1]
    U[0, N1:N4] = 0
    U[0, N4:N] = U[1, N4:N]
    U[M-1, 0:N2] = 0
    U[M-1, N2:N3] = 0
    U[M-1, N3:N] = 0

# Void function
# Reference
def set_boundary_V(V:np.ndarray):
    V[0:M, 0] = 0
    V[0:M, N-1] = 0
    V[0, 0:N1] = V[1, 0:N1]
    V[0, N1:N4] = 0
    V[0, N4:N] = V[1, N4:N]
    V[M-1, 0:N2] = 0
    V[M-1, N2:N3] = -1
    V[M-1, N3:N] = 0

# Void function
# Reference
def set_boundary_P(P:np.ndarray):
    P[0:M, 0] = P[0:M, 1]
    P[0:M, N-1] = P[0:M, N-2]
    P[0, 0:N1] = 0
    P[0, N1:N4] = P[1, N1:N4]
    P[0, N4:N] = 0
    P[M-1, 0:N2] = P[M-2, 0:N2]
    P[M-1, N2:N3] = 1
    P[M-1, N3:N] = P[M-2, N3:N]


start_x, end_x = (0, 1)
start_y, end_y = (0, 1)

N = 101
M = 101
dx = (end_x - start_x) / (N - 1)
dy = (end_y - start_y) / (M - 1)

dt = 0.0001
Re = 100
rho = 1
eps = 1e-6
stop_iteration = 5e4
eps_P = 1e-6
stop_iteration_P = 1e5

N1 = int(0.2 * N)
N2 = int(0.3 * N)
N3 = int(0.7 * N)
N4 = int(0.8 * N)

x = start_x + np.arange(start=0, stop=N) * dx
y = start_y + np.arange(start=0, stop=M) * dy
X, Y = np.meshgrid(x, y)

U_old = np.zeros((M, N))
U_new = np.zeros((M, N))

V_old = np.zeros((M, N))
V_new = np.zeros((M, N))

P_old = np.zeros((M, N))
P_new = np.zeros((M, N))

# Initial condition
# U(t=0, x, y) = 0
# V(t=0, x, y) = 0
# P(t=0, x, y) = 0
U_old[:, :] = 0
V_old[:, :] = 0
P_old[:, :] = 0
U_new[:, :] = 0
V_new[:, :] = 0
P_new[:, :] = 0

start_time = perf_counter()
# Boundary conditions
set_boundary_U(U=U_old)
set_boundary_V(V=V_old)

iteration = 0
maximum = 1
while maximum > eps and iteration < stop_iteration:
    # 1. Solve Burger's equation to find U*, V*
    U_new = Explicit_Burger(U_old, U_old, V_old, N, M, dx, dy, dt, 1/Re)
    V_new = Explicit_Burger(V_old, U_old, V_old, N, M, dx, dy, dt, 1/Re)
    set_boundary_U(U=U_new)
    set_boundary_V(V=V_new)

    # 2. Solve Poisson equation to find P
    set_boundary_P(P=P_old)
    set_boundary_P(P=P_new)

    iteration_P = 0
    maximum_P = 1
    while maximum_P > eps_P and iteration_P < stop_iteration_P:
        P_new = Jacobi_method(P_old, U_new, V_new, N, M, dx, dy, dt, rho)
        set_boundary_P(P=P_new)

        maximum_P = np.max(np.abs(P_new - P_old))

        P_old = P_new.copy()
        iteration_P += 1

    # Do not change
    # 3. Correction
    U_new[1:M-1, 1:N-1] = U_new[1:M-1, 1:N-1] - dt / rho*(
        P_new[1:M-1, 2:N] - P_new[1:M-1, 0:N-2]) \
            / (2*dx)  
    V_new[1:M-1, 1:N-1] = V_new[1:M-1, 1:N-1] - dt / rho*(
        P_new[2:M, 1:N-1] - P_new[0:M-2, 1:N-1]) \
            / (2*dy)

    set_boundary_U(U=U_new)
    set_boundary_V(V=V_new)

    max_U = np.max(np.abs(U_new - U_old))
    max_V = np.max(np.abs(V_new - V_old))
    maximum = max(max_U, max_V)

    U_old = U_new.copy()
    V_old = V_new.copy()

    if iteration % 20 == 0:
        print(f"{iteration}\t{maximum}\t{iteration_P}\t{maximum_P}")

    iteration += 1

end_time = perf_counter()

path = "Results"
if not os.path.exists(path):
    os.makedirs(path)

np.savetxt(f"{path}\\HW12_X_py.txt", X, fmt="%.6f", delimiter="\t")
np.savetxt(f"{path}\\HW12_Y_py.txt", Y, fmt="%.6f", delimiter="\t")
np.savetxt(f"{path}\\HW12_U_py.txt", U_new, fmt="%.6f", delimiter="\t")
np.savetxt(f"{path}\\HW12_V_py.txt", V_new, fmt="%.6f", delimiter="\t")
np.savetxt(f"{path}\\HW12_P_py.txt", P_new, fmt="%.6f", delimiter="\t")

with open("Results\\HW12.dat", "w") as file:
    file.write(f"VARIABLES = \"X\", \"Y\", \"U\", \"V\", \"P\"\n")
    file.write(f"ZONE I = {N}, J = {M}\n")
    
    for j in range(M):
        for i in range(N):
            file.write(f"{x[i]}\t{y[j]}\t{U_new[j, i]}\t" +
                       f"{V_new[j, i]}\t{P_new[j, i]}\n")
    
    print("Results are recorded")

winsound.Beep(750, 1000)

print("Results are recorded")
print("2D Navier-Stokes equation")
print("Explicit + Jacobi")
print(f"Calculation time: {end_time - start_time:.6f} seconds")
print(f"Number of iterations: {iteration}")
print(f"{N = }\n{M = }\n{dx = }\n{dy = }\n{dt = }")
print(f"{Re = }\n{rho = }")
print(f"{maximum = }\n{maximum_P = }")
print(f"U_absmax = {np.max(np.abs(U_new))}")
print(f"V_absmax = {np.max(np.abs(V_new))}")
print(f"P_absmax = {np.max(np.abs(P_new))}")

# Post Process
def draw_all(results, names, arrow=True):
    fig, ax = plt.subplots(2, 2)
    cnt = 0
    for row in range(2):
        for col in range(2):
            axs = ax[row, col]
            cf = axs.contourf(X, Y, results[cnt % 4])
            fig.colorbar(cf, ax=axs)
            axs.set_title(names[cnt % 4])
            if arrow:
                axs.streamplot(X, Y, U_new, V_new, color="black")
            axs.set_xlabel("x")
            axs.set_ylabel("y")
            cnt += 1
    plt.tight_layout()
    plt.show()
    
def draw_one(Z, name, arrow=False):
    fig, ax = plt.subplots()
    cf = ax.contourf(X, Y, Z)
    fig.colorbar(cf, ax=ax)
    if arrow:
        ax.streamplot(X, Y, U_new, V_new, color="black")
    plt.title(name)
    plt.xlabel("x")
    plt.ylabel("y")
    plt.show()


UV = (U_new**2 + V_new**2)**0.5
data = [U_new, V_new, P_new, UV]
name = ["U", "V", "P", "U + V"]
draw_all(data, name, arrow=True)

File: Week12/hw12_var2_jacobi.cpp
#include <iostream>
#include <cmath>
#include <chrono>
using namespace std::chrono;


// 2D Navier-Stokes equation
// Projection method
// Explicit + Jacobi
// Problem 1

// Initial condition
// U(t=0, x, y) = 0
// V(t=0, x, y) = 0

// Boundary conditions
// U(t, x=0, y) = 0
// U(t, x=1, y) = 0
// Uy(t, 0<x<0.2, y=0) = 0
// U(t, 0.2<x<0.8, y=0) = 0
// Uy(t, 0.8<x<1, y=0) = 0
// U(t, 0<x<0.3, y=1) = 0
// U(t, 0.3<x<0.7, y=1) = 0
// U(t, 0.7<x<1, y=1) = 0

// V(t, x=0, y) = 0
// V(t, x=1, y) = 0
// Vy(t, 0<x<0.2, y=0) = 0
// V(t, 0.2<x<0.8, y=0) = 0
// Vy(t, 0.8<x<1, y=0) = 0
// V(t, 0<x<0.3, y=1) = 0
// V(t, 0.3<x<0.7, y=1) = -1
// V(t, 0.7<x<1, y=1) = 0

// Px(t, x=0, y) = 0
// Px(t, x=1, y) = 0
// P(t, 0<x<0.2, y=0) = 0
// Py(t, 0.2<x<0.8, y=0) = 0
// P(t, 0.8<x<1, y=0) = 0
// Py(t, 0<x<0.3, y=1) = 0
// P(t, 0.3<x<0.7, y=1) = 1
// Py(t, 0.7<x<1, y=1) = 0

const double start_x = 0, end_x = 1;
const double start_y = 0, end_y = 1;

const int N = 101;
const int M = 101;

const double dx = (end_x - start_x) / (N - 1);
const double dy = (end_y - start_y) / (M - 1);

const int N1 = 0.2 * N;
const int N2 = 0.3 * N;
const int N3 = 0.7 * M;
const int N4 = 0.8 * M;

const double dt = 0.0001;
const double Re = 100;
const double nu = 1 / Re;
const double rho = 1;
const double eps = 1e-6;
const int stop_iteration = 3e4;
const double eps_P = 1e-6;
const int stop_iteration_P = 1e5;

double U_old[M][N], U_new[M][N];
double V_old[M][N], V_new[M][N];
double P_old[M][N], P_new[M][N];
double X[M][N], Y[M][N];

// Fills array with some value
void fill_array(double arr[M][N], double value=0){
    for(int j = 0; j < M; ++j){
        for(int i = 0; i < N; ++i){
            arr[j][i] = value;
        }
    }
}

// Returns maximum absolute difference of two arrays
double max_abs_diff(double A[M][N], double B[M][N]){
    double maximum = 0;
    for(int j = 0; j < M; ++j){
        for(int i = 0; i < N; ++i){
            if(maximum < fabs(A[j][i] - B[j][i])){
                maximum = fabs(A[j][i] - B[j][i]);
            }
        }
    }
    return maximum;
}

// copies values of array A to the array B
void copy_array(double A[M][N], double B[M][N]){
    for(int j = 0; j < M; ++j){
        for(int i = 0; i < N; ++i){
            B[j][i] = A[j][i];
        }
    }
}

void meshgrid(double X[M][N], double Y[M][N]){
    for(int j = 0; j < M; ++j){
        for(int i = 0; i < N; ++i){
            X[j][i] = start_x + i*dx;
            Y[j][i] = start_y + j*dy;
        }
    }
}

void savetxt(std::string path, double arr[M][N], 
                std::string fmt="%f", char delimeter=' '){
    const char * c = path.c_str();
    FILE *File;
    File = freopen(c, "w", stdout);
    std::string fmtwd = fmt + "%c";
    for(int j = 0; j < M; ++j){
        for(int i = 0; i < N; ++i){
            if(i != N-1){
                printf(fmt.c_str(), arr[j][i]);
                if(i != N-2){
                    printf("%c", delimeter);
                }
            }else{
                printf("\n");
            }
        }
    }
    fclose(File);
}

void Explicit_Burger(double S_new[M][N], double S_old[M][N], double U_old[M][N], double V_old[M][N], 
                        int N, int M, double dx, double dy, double dt, double nu){
    for(int j = 1; j < M-1; ++j){
        for(int i = 1; i < N-1; ++i){
            // central scheme
            S_new[j][i] = S_old[j][i] + dt*(
                - U_old[j][i]*(S_old[j][i+1] - S_old[j][i-1]) 
                                / (2*dx)
                - V_old[j][i]*(S_old[j+1][i] - S_old[j-1][i]) 
                                / (2*dy)
                + nu*(
                    (S_old[j][i+1] - 2*S_old[j][i] + S_old[j][i-1]) 
                        / (dx*dx) 
                    + (S_old[j+1][i] - 2*S_old[j][i] + S_old[j-1][i])
                        / (dy*dy))
            );
        }
    }
}

void Jacobi_Poisson(double P_new[M][N], double P_old[M][N], double U_new[M][N], double V_new[M][N], 
                      int N, int M, double dx, double dy, double dt, double rho){
    for(int j = 1; j < M-1; ++j){
        for(int i = 1; i < N-1; ++i){
            // central scheme
            P_new[j][i] = (
                dy*dy*(P_old[j][i+1] + P_old[j][i-1])
                + dx*dx*(P_old[j+1][i] + P_old[j-1][i])) 
                    / (2*(dx*dx + dy*dy))
                - dx*dx*dy*dy*rho
                    / (2*dt*(dx*dx + dy*dy))
                * ((U_new[j][i+1] - U_new[j][i-1]) 
                        / (2*dx) 
                    + (V_new[j+1][i] - V_new[j-1][i]) 
                        / (2*dy));
        }
    }           
}

// Boundary conditions
// U(t, x=0, y) = 0
// U(t, x=1, y) = 0
// Uy(t, 0<x<0.2, y=0) = 0
// U(t, 0.2<x<0.8, y=0) = 0
// Uy(t, 0.8<x<1, y=0) = 0
// U(t, 0<x<0.3, y=1) = 0
// U(t, 0.3<x<0.7, y=1) = 0
// U(t, 0.7<x<1, y=1) = 0
void set_boundary_U(double U[M][N]){
    for(int j = 0; j < M; ++j){
        U[j][0] = 0;
        U[j][N-1] = 0;
    }
    for(int i = 0; i < N1; ++i){
        U[0][i] = U[1][i];
    }
    for(int i = N1; i < N4; ++i){
        U[0][i] = 0;
    }
    for(int i = N4; i < N; ++i){
        U[0][i] = U[1][i];
    }
    for(int i = 0; i < N2; ++i){
        U[M-1][i] = 0;
    }
    for(int i = N2; i < N3; ++i){
        U[M-1][i] = 0;
    }
    for(int i = N3; i < N; ++i){
        U[M-1][i] = 0;
    }
}

// Boundary conditions
// V(t, x=0, y) = 0
// V(t, x=1, y) = 0
// Vy(t, 0<x<0.2, y=0) = 0
// V(t, 0.2<x<0.8, y=0) = 0
// Vy(t, 0.8<x<1, y=0) = 0
// V(t, 0<x<0.3, y=1) = 0
// V(t, 0.3<x<0.7, y=1) = -1
// V(t, 0.7<x<1, y=1) = 0
void set_boundary_V(double V[M][N]){
    for(int j = 0; j < M; ++j){
        V[j][0] = 0;
        V[j][N-1] = 0;
    }
    for(int i = 0; i < N1; ++i){
        V[0][i] = V[1][i];
    }
    for(int i = N1; i < N4; ++i){
        V[0][i] = 0;
    }
    for(int i = N4; i < N; ++i){
        V[0][i] = V[1][i];
    }
    for(int i = 0; i < N2; ++i){
        V[M-1][i] = 0;
    }
    for(int i = N2; i < N3; ++i){
        V[M-1][i] = -1;
    }
    for(int i = N3; i < N; ++i){
        V[M-1][i] = 0;
    }
}

// Px(t, x=0, y) = 0
// Px(t, x=1, y) = 0
// P(t, 0<x<0.2, y=0) = 0
// Py(t, 0.2<x<0.8, y=0) = 0
// P(t, 0.8<x<1, y=0) = 0
// Py(t, 0<x<0.3, y=1) = 0
// P(t, 0.3<x<0.7, y=1) = 1
// Py(t, 0.7<x<1, y=1) = 0
void set_boundary_P(double P[M][N]){
    for(int j = 0; j < M; ++j){
        P[j][0] = P[j][1];
        P[j][N-1] = P[j][N-2];
    }
    for(int i = 0; i < N1; ++i){
        P[0][i] = 0;
    }
    for(int i = N1; i < N4; ++i){
        P[0][i] = P[1][i];
    }
    for(int i = N4; i < N; ++i){
        P[0][i] = 0;
    }
    for(int i = 0; i < N2; ++i){
        P[M-1][i] = P[M-2][i];
    }
    for(int i = N2; i < N3; ++i){
        P[M-1][i] = 1;
    }
    for(int i = N3; i < N; ++i){
        P[M-1][i] = P[M-2][i];
    }
}


int main(){
    // Filling grid
    meshgrid(X, Y);

    // Initial condition
    fill_array(U_old, 0);
    fill_array(U_new, 0);
    fill_array(V_old, 0);
    fill_array(V_new, 0);
    fill_array(P_old, 0);
    fill_array(P_new, 0);

    auto start = high_resolution_clock::now();

    // Boundary conditions
    set_boundary_U(U_old);
    set_boundary_V(V_old);

    int iteration = 0;
    double maximum = 0;
    int iteration_P = 0;
    double maximum_P = 0;
    do{
        Explicit_Burger(U_new, U_old, U_old, V_old, N, M, dx, dy, dt, nu);
        Explicit_Burger(V_new, V_old, U_old, V_old, N, M, dx, dy, dt, nu);
        set_boundary_U(U_new);
        set_boundary_V(V_new);

        set_boundary_P(P_old);
        set_boundary_P(P_new);

        iteration_P = 0;
        maximum_P = 0;
        do{
            Jacobi_Poisson(P_new, P_old, U_new, V_new, N, M, dx, dy, dt, rho);
            
            set_boundary_P(P_new);            
            maximum_P = max_abs_diff(P_new, P_old);
            copy_array(P_new, P_old);
            iteration_P++;
        }while(maximum_P > eps_P and iteration_P < stop_iteration_P);

        // Correct U, V
        for(int j = 1; j < M-1; ++j){
            for(int i = 1; i < N-1; ++i){
                U_new[j][i] = U_new[j][i] - dt / rho * (P_new[j][i+1] - P_new[j][i-1]) / (2*dx);
                V_new[j][i] = V_new[j][i] - dt / rho * (P_new[j+1][i] - P_new[j-1][i]) / (2*dy);
            }
        }
        
        set_boundary_U(U_new);
        set_boundary_V(V_new);

        double maximum_U = max_abs_diff(U_new, U_old);
        double maximum_V = max_abs_diff(V_new, V_old);
        maximum = std::fmax(maximum_U, maximum_V);
        
        copy_array(U_new, U_old);
        copy_array(V_new, V_old);
        
        if(iteration % 2 == 0){
            std::cout << iteration << "\t" << maximum << "\t" 
            << iteration_P << "\t" << maximum_P << "\n";
        }

        iteration++;
    }while(maximum > eps and iteration < stop_iteration);

    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start);
    printf("calculating time: %.6f seconds\n", duration.count() / 1e6);

    std::cout << "Explicit method\n";
    std::cout << "Number of iterations = " << iteration << std::endl;
    std::cout << "N = " << N << std::endl;
    std::cout << "M = " << M << std::endl;
    std::cout << "dx = " << dx << std::endl;
    std::cout << "dy = " << dy << std::endl;
    std::cout << "dt = " << dt << std::endl;
    std::cout << "Re = " << Re << std::endl;
    std::cout << "rho = " << rho << std::endl;
    std::cout << "nu = " << nu << std::endl;
    std::cout << "eps = " << eps << std::endl;
    std::cout << "Maximum = " << maximum << std::endl;

    savetxt("Results\\HW12_X_cpp.txt", X, "%.6f", '\t');
    savetxt("Results\\HW12_Y_cpp.txt", Y, "%.6f", '\t');
    savetxt("Results\\HW12_U_cpp.txt", U_new, "%.6f", '\t');
    savetxt("Results\\HW12_V_cpp.txt", V_new, "%.6f", '\t');
    savetxt("Results\\HW12_P_cpp.txt", P_new, "%.6f", '\t');
    
    std::cout << "Results are recorded" << std::endl;

    freopen("Results\\HW12_cpp.dat", "w", stdout);
    std::cout << "VARIABLES = \"X\", \"Y\", \"U\", \"V\", \"P\"" << std::endl;
    std::cout << "ZONE I = " << N << ", J = " << M << std::endl;

    for(int j = 0; j < M; ++j){
        for(int i = 0; i < N; ++i){
            std::cout << X[i] << "\t" << Y[j] << "\t" << U_new[j][i] 
                << "\t" << V_new[j][i] << "\t" << P_new[j][i] << std::endl;
        }
    }

    return 0;
}
